Added tests for template generation

santacodes opened the pull request:
# Description 
- Added basic tests for template generation using `pytest-cookies`.
- Added `cookiecutter` as a dependency and setup `___init.py__` for the src layout.

Saransh-cpp reviewed (changes requested):
Thanks, @santacodes! Looks good to me. Could you please add a minimal CI job to check if the tests are working fine?

arjxn-py commented on tests/test_project_generation.py:
```diff
@@ -0,0 +1,31 @@
+    return template
+
+
+def test_bake_custom_project(cookies, custom_template):
```
```suggestion
def test_bake_custom_project(cookies: Cookies, custom_template):
```

Don't want to make the work harder but can we also take care of typing a bit from the start just in case you're familiar with it?

arjxn-py commented on tests/test_project_generation.py:
```diff
@@ -0,0 +1,31 @@
+import pybamm_cookiecutter as m
+import pytest
```
```suggestion
import pytest
from pytest_cookies.plugin import Cookies
```

Related to the other typing comment.

arjxn-py reviewed (commented):
Thanks @santacodes, Good Start I must admit :) 
Some comments below to propose some improvements

santacodes commented on tests/test_project_generation.py:
```diff
@@ -0,0 +1,31 @@
+    assert m.__version__
+
+@pytest.fixture
+def custom_template(tmpdir):
```
Sorry for the late reply, I was travelling, I would also mention `tmp_path` returns [py.path](https://py.readthedocs.io/en/latest/path.html) which also quotes 
> Note: The ‘py’ library is in “maintenance mode” and so is not recommended for new projects. Please check out [pathlib](https://docs.python.org/3/library/pathlib.html) or [pathlib2](https://pypi.org/project/pathlib2/) for path operations.

So I am not sure which one shall I proceed with.

agriyakhetarpal replied:
I think `tmp_path` should be better, then. It's easier to work with `pathlib.Path`

santacodes commented:
> Thanks, @santacodes! Looks good to me. Could you please add a minimal CI job to check if the tests are working fine?

Sorry for the delay @Saransh-cpp, I was travelling. I added CI for the tests, created a nox session for testing the template generation, and tested it on my branch. Could you please review and look at it? Also, I added uv to the nox backend, which seems to be faster.

santacodes commented on tests/test_project_generation.py:
```diff
@@ -0,0 +1,31 @@
+    assert result.exit_code == 0
+    assert result.exception is None
+
+    assert result.project_path.name == "pybamm_cookie"
```
Added them in the latest commit, do they seem fine or would we be better with the usual `try-except` blocks?

arjxn-py replied:
It's perfect, we don't need try-expect blocks for this.

Saransh-cpp commented on noxfile.py:
```diff
@@ -18,3 +19,12 @@ def build_docs(session: nox.Session) -> None:
+def run_template_generation(session):
+    """Run the tests tests for testing template generation"""
+    session.install("setuptools", silent=False)
+    session.install("pytest", silent=False)
+    session.install("pytest-cookies", silent=False)
+    session.install("-e", ".", silent=False)
```
```suggestion
    session.install("-e", ".[dev]", silent=False)
```

Saransh-cpp commented on src/pybamm_cookiecutter/__init__.py:
```diff
@@ -5,6 +5,11 @@
-__all__ : tuple[str] = ("__version__",)
+__all__ : list[str] = [
+    "__version__",
+    "pybamm",
```
pybamm should not be importable through this module
```suggestion
```

Saransh-cpp commented on .github/test_on_push.yml:
```diff
@@ -0,0 +1,53 @@
+name: pybamm_cookiecutter
+
+on:
+  workflow_dispatch:
+  pull_request:
```
Should we also run it with `push: [main]`?

santacodes replied:
added it in the latest commit

Saransh-cpp reviewed (changes requested):
Thanks, @santacodes! Few minor changes below -

Saransh-cpp commented:
TODO:
- [ ] Update branch protection rule to include the new CI (status checks must pass before merging)

santacodes commented:
> TODO:
> 
> * [ ]  Update branch protection rule to include the new CI (status checks must pass before merging)

![branchrule](https://github.com/pybamm-team/pybamm-cookiecutter/assets/52504160/4a3dacaa-d79f-4e3d-8264-84ee24242c86)

I think I can only create a branch rule allowing merges only after status checks when the status checks themselves have been merged to the main branch.

Saransh-cpp commented:
Hi, don't worry about the rules. I'll set them up.

Saransh-cpp reviewed (approved):
This looks good to me, thanks @santacodes! I'll wait for @agriyakhetarpal and @arjxn-py's reviews.

arjxn-py commented:
Hi, I'll review this in a couple of hours.

arjxn-py commented on tests/test_project_generation.py:
```diff
@@ -0,0 +1,30 @@
+    Testing if the projects exists in the tempdir
+    """
+    result = cookies.bake(template=str(custom_template))
+    assert result.exit_code == 0, "Exited with code 0"
```
```suggestion
    assert result.exit_code == 0, f"Exited with code {result.exit_code}, expected 0"
```
This should be more helpful, correct me if needed :)

arjxn-py commented on tests/test_project_generation.py:
```diff
@@ -0,0 +1,30 @@
+    assert result.exit_code == 0, "Exited with code 0"
+    assert result.exception is None, result.exception
+    assert result.project_path.name == "pybamm_cookie"
+    assert result.project_path.is_dir(), "Project directory not found"
```
```suggestion
    assert result.project_path.is_dir(), f"Project directory {result.project_path} not found"
```

arjxn-py reviewed (commented):
Looks much better @santacodes 🙂, Some comments below for minor improvements, feel free to ping me again.
Before we merge, we would want that the workflow starts running on this same PR.
Plus, I really appreciate that we're also adapting to `uv` in this. 
xref [PyBaMM#3825](https://github.com/pybamm-team/PyBaMM/issues/3825)

agriyakhetarpal commented on .github/test_on_push.yml:
```diff
@@ -0,0 +1,55 @@
+name: pybamm_cookiecutter
```
```suggestion
name: Test template generation
```

agriyakhetarpal commented on .github/test_on_push.yml:
```diff
@@ -0,0 +1,55 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macos-12, macos-14, windows-latest]
```
```suggestion
        os: [ubuntu-latest, macos-13, macos-14, windows-latest]
```

agriyakhetarpal commented on .github/test_on_push.yml:
```diff
@@ -0,0 +1,55 @@
+    name:
+      Template Tests (${{ matrix.os }} / Python ${{ matrix.python-version }})
+    steps:
+      - name: Checkout pybamm_cookiecutter
```
```suggestion
      - name: Checkout pybamm-cookiecutter
```

agriyakhetarpal commented on .github/test_on_push.yml:
```diff
@@ -0,0 +1,55 @@
+          python-version: ${{ matrix.python-version }}
+          cache: 'pip'
+
+      - name: Set up uv
+        uses: yezz123/setup-uv@v4
+        with:
+          uv-venv: ".venv"
```
Question: is it possible to cache `uv` through this action? I haven't explored it. We should cache either `pip` or `uv`, since both of them store cached wheels in difference places, IIRC.

agriyakhetarpal replied:
Yes, makes sense – let's look at `uv`'s caching later during the project. As an aside, I did a quick run for `uv` in my PyBaMM repository locally and it works quite well now, saving up a minute of time and more.

agriyakhetarpal commented on pyproject.toml:
```diff
@@ -30,14 +30,15 @@ classifiers = [
-dependencies = ["pybamm"]
+dependencies = ["pybamm", "cookiecutter"]
 
 [project.optional-dependencies]
```
I would suggest changing `nox` in this table to `nox[uv]`, otherwise it will always use `virtualenv`.

agriyakhetarpal commented on pyproject.toml:
```diff
@@ -30,14 +30,15 @@ classifiers = [
 dev = [
   "pytest >=6",
   "pytest-cov >=3",
   "nox",
```
```suggestion
  "nox[uv]",
```

agriyakhetarpal commented on noxfile.py:
```diff
@@ -18,3 +19,10 @@ def build_docs(session: nox.Session) -> None:
+
+@nox.session(name="test-generation")
+def run_template_generation(session):
+    """Run the tests tests for testing template generation"""
```
```suggestion
    """Run the tests for testing template generation"""
```

agriyakhetarpal commented:
P.S. I'm sorry for the late review! I have been a bit out of touch, plus I noticed that Saransh and Arjun had already put in their reviews. I hope this is helpful.

arjxn-py reviewed (approved):
Thanks @santacodes, the tests look perfect to me now.

arjxn-py commented:
Before we merge, I'd just want to confirm if we can safely resolve the conversation about `uv` caching plus can you maybe link the CI run on the latest commit from your fork here?

agriyakhetarpal commented on .github/test_on_push.yml:
```diff
@@ -0,0 +1,55 @@
+          uv-venv: ".venv"
+
+      - name: Install nox
+        run: uv pip install nox
```
```suggestion
        run: uv pip install nox[uv]
```

agriyakhetarpal reviewed (dismissed):
This looks good to me as long as the tests pass – resolved the previous conversation. Thanks for adding this, @santacodes!

xrefs for `uv`'s caching to look at later: https://github.com/astral-sh/uv/issues/2231, https://github.com/yezz123/setup-uv/issues/25, https://github.com/actions/setup-python/issues/822, https://github.com/actions/setup-python/pull/818

agriyakhetarpal commented:
(I think we should disable the setting that dismisses reviews on further commits)

arjxn-py commented:
I'm merging this as @santacodes needs to start working on entrypoints as well.

arjxn-py commented:
> (I think we should disable the setting that dismisses reviews on further commits)

Yes, I was going to mention that reviews were getting dismissed automatically
